File: composeApp/src/androidMain/kotlin/com/app/starleet/viewmodel/LactateViewModel.kt
package com.app.starleet.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.app.starleet.intent.LactateIntent
import com.app.starleet.roomdb.LactateRepository
import com.app.starleet.states.LactateState
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class LactateViewModel(
    private val repository: LactateRepository,
    application: Application
) : AndroidViewModel(application) {

    private val _state = MutableStateFlow(LactateState())
    val state: StateFlow<LactateState> = _state

    init {
        processIntent(LactateIntent.LoadLactate)
    }

    fun processIntent(intent: LactateIntent) {
        when (intent) {

            is LactateIntent.LoadLactate -> loadLactate()

            is LactateIntent.AddScan -> addScan(
                intent.value,
                intent.sweat,
                intent.change
            )

            is LactateIntent.SelectHistoryDate -> {
                _state.update { it.copy(selectedDate = intent.dateMillis) }
            }
        }
    }

    private fun loadLactate() {

        viewModelScope.launch {

            repository.allScans.collect { allData ->

                val selectedDate = _state.value.selectedDate

                val filteredHistory = if (selectedDate != null) {

                    val start = getStartOfDay(selectedDate)
                    val end = start + (24 * 60 * 60 * 1000)

                    allData.filter {
                        it.timestamp in start until end
                    }

                } else {
                    allData
                }

                _state.update {
                    it.copy(
                        history = filteredHistory,
                        trend = allData.takeLast(7)
                    )
                }
            }
        }

        viewModelScope.launch {
            val latest = repository.getLatest()
            _state.update { it.copy(current = latest) }
        }
    }

    private fun addScan(
        value: Double,
        sweat: String,
        change: Double
    ) {
        viewModelScope.launch {


            val randomValue = (value + (0..10).random() / 10.0)
                .coerceIn(0.0, 4.0)
                .roundTo1Decimal()

            val randomChange = ((-5..5).random() / 10.0)
                .roundTo1Decimal()

            repository.insertScan(
                value = randomValue,
                sweat = sweat,
                change = randomChange
            )


            val latest = repository.getLatest()

            _state.update {
                it.copy(current = latest)
            }
        }
    }

    private fun getStartOfDay(time: Long): Long {
        val cal = java.util.Calendar.getInstance()
        cal.timeInMillis = time
        cal.set(java.util.Calendar.HOUR_OF_DAY, 0)
        cal.set(java.util.Calendar.MINUTE, 0)
        cal.set(java.util.Calendar.SECOND, 0)
        cal.set(java.util.Calendar.MILLISECOND, 0)
        return cal.timeInMillis
    }

    fun Double.roundTo1Decimal(): Double {
        return String.format("%.1f", this).toDouble()
    }
}

File: iosApp/iosApp/ViewModel/LactateViewModel.swift
//
//  LactateViewModel.swift
//  iosApp
//
//  Created by KBS on 3/3/26.
//

import SwiftUI
import SwiftData
import AVFoundation

@MainActor
class LactateViewModel: ObservableObject {

    @Published private(set) var state = LactateState()

    private let repository: LactateRepository

    init(repository: LactateRepository) {
        self.repository = repository
        processIntent(.loadLactate)
    }

    func processIntent(_ intent: LactateIntent) {
        switch intent {

        case .loadLactate:
            loadLactate()

        case .addScan(let value, let sweat, let change):
            addScan(value: value, sweat: sweat, change: change)

        case .selectHistoryDate(let date):
            state.selectedDate = date
            loadLactate()
        }
    }

    private func loadLactate() {

        let allData = repository.getAllScans()

        if let selectedDate = state.selectedDate {

            let start = Calendar.current.startOfDay(for: selectedDate)
            let end = Calendar.current.date(byAdding: .day, value: 1, to: start)!

            state.history = allData.filter {
                $0.timestamp >= start && $0.timestamp < end
            }

        } else {
            state.history = allData
        }

        state.trend = Array(allData.prefix(7))
        state.current = repository.getLatestScan()
    }

    private func addScan(value: Double, sweat: String, change: Double) {

        let randomValue = (value + Double.random(in: 0...1))
            .clamped(to: 0...4)
            .rounded(toPlaces: 1)

        let randomChange = Double.random(in: -0.5...0.5)
            .rounded(toPlaces: 1)

        repository.insertScan(
            value: randomValue,
            sweat: sweat,
            change: randomChange
        )

        loadLactate()
    }
}
extension Double {
    func rounded(toPlaces places: Int) -> Double {
        let divisor = pow(10.0, Double(places))
        return (self * divisor).rounded() / divisor
    }

    func clamped(to range: ClosedRange<Double>) -> Double {
        min(max(self, range.lowerBound), range.upperBound)
    }
}
